src/Core/Form/ChoiceProvider/LanguageByIdChoiceProvider.php

```diff
@@ -35,16 +34,16 @@
 final class LanguageByIdChoiceProvider implements FormChoiceProviderInterface
 {
     /**
-     * @var LanguageDataProvider
+     * @var array
      */
-    private $languageDataProvider;
+    private $languages;
 
     /**
-     * @param LanguageDataProvider $languageDataProvider
+     * @param array $languages
      */
-    public function __construct(LanguageDataProvider $languageDataProvider)
+    public function __construct(array $languages)
```

Comment:
Unfortunately, yes, because `FormChoiceProviderInterface` does not allow to pass options. Maybe you can use a Decorator to make the 2 related and avoid duplicate code ?

Can you add a `@todo` statement in this new class, stating that is is very close to LanguageIdChoiceProvider and suggesting a refactor to merge these 2 ?

We could use a setter/configuratir to make the FormChoiceProvider stateful and obtain the desired behavior (something like `allowOnlyActiveLanguageChoices()` but that would seem wrong 🤔 ?

Reply:
> Unfortunately, yes, because `FormChoiceProviderInterface` does not allow to pass options. Maybe you can use a Decorator to make the 2 related and avoid duplicate code ?

IMO decorating `LanguageByIdChoiceProvider` would be redundant in this case, as the code in decorator would be nearly similar to code in this new `LanguageChoiceProvider`. I'd need to inject all languages to the decorator, then check if the language exists in the choices array already and only if it doesn't exist -insert it there. Or ignore the code from the class that's being decorated and just use all language choices, which makes more sense, but neglects the idea behind decorator pattern.




> Can you add a `@todo` statement in this new class, stating that is is very close to LanguageIdChoiceProvider and suggesting a refactor to merge these 2 ?

Added this `@todo`.
 
> We could use a setter/configuratir to make the FormChoiceProvider stateful and obtain the desired behavior (something like `allowOnlyActiveLanguageChoices()` but that would seem wrong  ?

I agree, it seems wrong :confused: IMO even if the code is kind of duplicated, it's better to have two similar classes with an intention to merge them into one at some point, than to have workarounds in code.